src/Features/VisualBasic/Portable/Completion/CompletionProviders/SymbolCompletionProvider.vb

```diff
@@ -78,22 +78,65 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.Providers
             Return MyBase.GetFilterText(symbol, displayText, context)
         End Function
 
-        Private Shared s_importDirectiveRules As CompletionItemRules =
-            CompletionItemRules.Create(commitCharacterRules:=ImmutableArray.Create(CharacterSetModificationRule.Create(CharacterSetModificationKind.Replace, "."c)))
-        Private Shared s_importDirectiveRules_preselect As CompletionItemRules =
-            s_importDirectiveRules.WithSelectionBehavior(CompletionItemSelectionBehavior.SoftSelection)
-
-        ' '(' should not filter the completion list, even though it's in generic items like IList(Of...)
-        Private Shared ReadOnly s_itemRules As CompletionItemRules = CompletionItemRules.Default.
-            WithFilterCharacterRule(CharacterSetModificationRule.Create(CharacterSetModificationKind.Remove, "("c)).
-            WithCommitCharacterRule(CharacterSetModificationRule.Create(CharacterSetModificationKind.Add, "("c))
-
-        Private Shared ReadOnly s_itemRules_preselect As CompletionItemRules = s_itemRules.WithSelectionBehavior(CompletionItemSelectionBehavior.SoftSelection)
-
-        Protected Overrides Function GetCompletionItemRules(symbols As List(Of ISymbol), context As SyntaxContext, preselect As Boolean) As CompletionItemRules
-            Return If(context.IsInImportsDirective,
-                If(preselect, s_importDirectiveRules_preselect, s_importDirectiveRules),
-                If(preselect, s_itemRules_preselect, s_itemRules))
+        Private Shared cachedRules As Dictionary(Of ValueTuple(Of Boolean, Boolean, Boolean), CompletionItemRules) =
+            InitCachedRules()
+
+        Private Shared Function InitCachedRules() As Dictionary(Of ValueTuple(Of Boolean, Boolean, Boolean), CompletionItemRules)
+
+            Dim result = New Dictionary(Of ValueTuple(Of Boolean, Boolean, Boolean), CompletionItemRules)()
+
+            For importDirective = 0 To 1
+                For preselect = 0 To 1
+                    For tuple = 0 To 1
+                        If importDirective = 1 AndAlso tuple = 1 Then
+                            Continue For
```

Comment:
@rchande Readability was not the motivation here. There used to be 4 rules (2 x 2 combinations). Now there are 2 x 2 x 2. Except that one of those is meaningless and Cyrus is making fun of that optimization ;-)